### ISSUE TITLE
Teehistorian Replayer

### ISSUE THREAD

**Issue by def-** [ddnet/ddnet#2937]
I haven't really started yet, if someone wants to help tell me:

- Replay modus on server that opens a teehistorian file as inputs
- Should run faster than realtime
- Serverside demo recording enabled
- Need a way to tell get partial demos efficiently, some teehistorian files are multi-GB when server runs for days at a time with full players, maybe split by time or search for player names. maybe find a finish, search back to where players went through start line, not sure if this can be implemented.

The end goal is to get demos out for the top 1 ranks of all maps on DDNet. Even better would be a teewebs based viewer. (Someone on Discord recently got teewebs by eeeee building again, we can check who it was if there is interest in making it into a demo watcher.)

**Comment by Zwelf** [comment]
I wrote a [teehistorian parser](https://gitlab.com/zwelf/teehistorian/) with a C/C++ API. I made the C/C++ bindings so that it can be easily integrated into the DDNet-Server for replaying teehistorian files.

**Comment by def-** [comment]
Very cool. Are you planning to keep working on this?

**Comment by Zwelf** [comment]
Yes I do plan to continue working on the lib and will try to integrate it into DDNet-Server. I don't know yet where to start integrating the teehistorian player in the DDNet source code (and I tried to find a starting point once). Therefore I would really appreciate help regarding the beginning of the integration.

**Comment by def-** [comment]
@heinrich5991 Do you have some ideas?

**Comment by def-** [comment]
@Zwelf Is it makeing any progress? Do you still need help with the beginning of integration?

**Comment by Learath2** [comment]
We also should check how well the serverside demo recording works. I had to mess around a lot with it while getting the bridge code in, I might have missed a couple messages.

**Comment by Zwelf** [comment]
No, I don't make progress right now and I don't find time to work on this either. I would be happy if someone else wants to pick this issue up. I can help with using my library if needed.
I might look back at this issue in the future if no one else does. But I don't want to give any guarantees, that this will happen soon. Sorry for not responding so far.

**Comment by MagicUnicornScReeNy** [comment]
### Top Rank demos should definitely not go public!
Obviously you might just say that im mad a lot my skip would get leaked, but that's not the whole point here. I love checking Maps for Skips and don't want get spoiled or anything. When i see a map with an insanely good rank 1, i just wanna theory craft the pathing for it and beat it. Figuring out skips and pathings is a huge aspect of the game, which shouldn't get obsolete. This is part of what makes this game so special. Its not your typical 95% mechanical skills only required Game like the often as example used Trackmania. This Game is about Skill, Brain and Teamwork and you should have all of it to rule the leader boards (except u coradax then u dont need teamwork).

The work done here isn't obsolete either, it can be used to make verifying ranks much easier.

**Comment by Learath2** [comment]
I disagree with theorycrafting odd skips that will be patched out in a couple hours  being a HUGE part of the game. The patterns of maps played by the community also indicate that most of the people are here for the extremely high skill cap. 

I think the theorycrafting belongs in the maps specifically made as puzzles breaking from the pattern (as we are all very familiar with most types of parts). 

If you don't want to be spoiled about maps, then don't watch them. Unbeatable ranks on leaderboards just reeks of elitism and gatekeeping.

**Comment by Zwelf** [comment]
The best way forward for me to replay would be to make the game physics compile stand alone as a library. This would be mostly `CGameContext`. Not sure how easy that would be. I'd be happy if someone else would like to work on that! Then I could use my teehistorian library+replayer with the ddnet physics.

**Comment by def-** [comment]
@ChillerDragon You already started some work on isolating parts of client. Do you think something similar would also be possible for server/CGameContext?

**Comment by ChillerDragon** [comment]
> @ChillerDragon You already started some work on isolating parts of client. Do you think something similar would also be possible for server/CGameContext?

@Learath2 knows better for sure. But I see no issue with running a full server with high tickspeed.

**Comment by Zwelf** [comment]
> But I see no issue with running a full server with high tickspeed.

Yeah modifying the server works too. If you want to work on this, it would be great as well. You would probably need to replace the networking code somehow to use teehistorian instead. But I've also written a teehistorian verifier in Rust (I'll publish it soon, no worries) based on https://gitlab.com/Zwelf/teehistorian, which could get a CGameContext and call the functions in the appropriate order. So for me, creating a stand alone CGameContext library to replay teehistorian files with the ddnet physics would be the least additional effort if possible.
